Sq1 eliana/s07 listagem treinos estudante

ElianaMT opened the pull request:
Revisao backend finalizado

ElianaMT committed:
feat:adiciona credenciais

ElianaMT committed:
feat:cria rota, controller e metodo

O metodo lista de forma geral os workouts

ElianaMT committed:
feat: workouts clasificados por student_id

ElianaMT committed:
feat:workouts ordenados pelo dia da semana

ElianaMT committed:
feat:adiciona clasificacao por created_at do registro de workouts

ElianaMT committed:
feat:adiciona filtro por student_id e muda rota

HTTP GET no path /api/students/:id/workouts

ElianaMT committed:
feat: mostra student_name no array

ElianaMT committed:
feat:join con tablea exercises para mostrar a description do exercise

ElianaMT committed:
feat:implementa formattigWorkouts

CarolineSampaio commented on app/Http/Controllers/WorkoutController.php:
```diff
@@ -17,10 +17,43 @@ public function index()
 
+    public function workoutsByStudent($studentId)
+    {
+        try {
```
O bloco try catch pode ser removido, pois eu fiz uma configuração no projeto todo que já captura a exceção.

CarolineSampaio commented on app/Http/Controllers/WorkoutController.php:
```diff
@@ -17,10 +17,43 @@ public function index()
+    public function workoutsByStudent($studentId)
+    {
+        try {
+            $workouts = Workout::select('workouts.*', 'students.name', 'exercises.description')
```
Apesa de estar funcionando, aqui podiamos usar códigos do laravel 10 ao invés da query pura.

Ficará assim:
  $workouts = Workout::with(['student', 'exercise'])
              ->where('student_id', $studentId)
              ->orderBy('created_at')
              ->get();

```php
$workouts = Workout::with(['student', 'exercise'])
                      ->where('student_id', $studentId)
                      ->orderBy('created_at')
                      ->get();
```
            
Bem menor né? Para que isso funcione precisamos fazer o relacionamento la no model. 

```php 
     public function student()
    {
        return $this->belongsTo(Student::class, 'student_id');
    }

    public function exercise()
    {
        return $this->belongsTo(Exercise::class, 'exercise_id');
    }
```

CarolineSampaio commented on app/Http/Controllers/WorkoutController.php:
```diff
@@ -17,10 +17,43 @@ public function index()
+                $formattingWorkouts['student_id'] = $workout->student_id;
+                $formattingWorkouts['name'] = $workout->name;
+                
+                $day = $workout->day;
```
Esse workout->day poderia ser colocado diretamente:

```php
$formattingWorkouts['workouts'][$workout->day][] = [
....
```

CarolineSampaio commented on app/Http/Controllers/WorkoutController.php:
```diff
@@ -17,10 +17,43 @@ public function index()
+                ];
+            }
+    
+            return response()->json($formattingWorkouts);
```
O professor disponibilizou para nós uma traits a HttpResponses, que já está até importada no controller, bastaria alterar o returno para o seguinte código:

```php
 return $this->response('Listagem de treinos realizada com sucesso', Response::HTTP_OK, $formattingWorkouts);
```

Primeiro parâmetro -> mensagem, que poderia ser vazia também;
Segundo parâmetro -> status code;
Terceiro parâmetro-> array

CarolineSampaio commented on routes/api.php:
```diff
@@ -19,6 +19,7 @@
 Route::middleware('auth:sanctum')->group(function () {
     Route::get('students', [StudentController::class, 'index']);
     Route::get('workouts', [WorkoutController::class, 'index']);
+    Route::get('students/{id}/workouts', [WorkoutController::class, 'workoutsByStudent']);
```
Aqui na rota ficou faltando o ability para proteger a aplicação e garantir que apenas alunos possam ter acesso à essa rota.

```php
    Route::get('students/{id}/workouts', [WorkoutController::class, 'workoutsByStudent'])->middleware(['ability:get-workout']);
```

Já coloquei a ability correta

CarolineSampaio commented on .env.example:
```diff
@@ -11,7 +11,7 @@ LOG_LEVEL=debug
 DB_HOST=127.0.0.1
 DB_PORT=5432
-DB_DATABASE=academia_api
+DB_DATABASE=academiaf_api
```
Esse arquivo de configuração comum de todos, então não pode fazer essa alteração. Se fizer em seu computador, não poderá fazer o commit pois pode comprometer os colegas;

CarolineSampaio reviewed (changes requested):
Boa noite, Eliana.

Fiz alguns comentários a respeito do código com sugestões de melhorias. Além deles, eu recomendária a implementação do conceito de service, repository e interface.

Fico à disposição para responder suas dúvidas e também para fazemos uma ligação para conversar e programar junto

ElianaMT committed:
refactor:revert.env.example

ElianaMT committed:
Merge branch 'features/admin' of https://github.com/DEVinHouse-Zucchetti/M3P-BackEnd into sq1-eliana/s07-listagem-treinos-estudante

ElianaMT committed:
refactor: atualiza middleware ability

ElianaMT committed:
refactor:muda resposta return

ElianaMT committed:
Merge branch 'features/admin' into eliana/revisao/back

ElianaMT committed:
Merge remote-tracking branch 'origin/sq1-marcel/s01-login' into eliana/revisao/back

ElianaMT committed:
refactor: adiciona authentication de user no Controller Workout

ElianaMT committed:
feat:cria arquivo StudentWorkoutTest

ElianaMT committed:
feat:adiciona test que verifica status code =200

ElianaMT committed:
feat: teste 2 back estrutura do array de resposta

ElianaMT committed:
feta:adiciona 2 testes

 test_student_can_list_their_own_workouts
test_student_can_not_list_workouts_from_another_student

ElianaMT committed:
Merge branch 'features/admin' into eliana/revisao/test

ElianaMT committed:
feat:atualiza o codigo

 Apaga try, catch
Muda foreach para que o studante autenticado somente possa ver seus treinos

ElianaMT committed:
feat: adiciona a la rota ability

ElianaMT committed:
Merge branch 'eliana/revisao/test' into sq1-eliana/s07-listagem-treinos-estudante

CarolineSampaio commented on tests/Unit/StudentWorkoutTest.php:
```diff
@@ -0,0 +1,78 @@
+
+class StudentWorkoutTest extends TestCase
+{
+    use DatabaseTransactions;
```
Linhas 11 a 58 podem ser removidas pois duplicam o que já existe nos últimos dois códigos.

O Data base transactions já foi aplicado globalmente também :D

CarolineSampaio commented on tests/Unit/StudentWorkoutTest.php:
```diff
@@ -0,0 +1,78 @@
+        $token = $student->createToken('@academia', ['get-workout'])->plainTextToken;
+
+        $response = $this->withHeader('Authorization', 'Bearer ' . $token)->get('/api/students/' . $student->id . '/workouts');
+        $response->assertStatus(200);
```
Além de dar um assert no status podemos colocar a estrutura esperada aqui:

```php
$response->assertStatus(200)->assertJsonStructure([
            'message',
            'status',
            'data' => [
                'student_id',
                'name',
                'workouts' => [
                    'SEGUNDA',
                    'TERÇA',
                    'QUARTA',
                    'QUINTA',
                    'SEXTA',
                    'SÁBADO',
                    'DOMINGO',
                ],
            ],
        ]);
```

ElianaMT committed:
refactor: adiciona middleware na rota

ElianaMT committed:
Merge branch 'eliana/revisao/back_1' into sq1-eliana/s07-listagem-treinos-estudante

ElianaMT committed:
Merge branch 'features/admin' into eliana/revisao/back_3

ElianaMT committed:
feat:actualiza branch com admin

ElianaMT committed:
refactor: join tabela workouts con tabela users_students

ElianaMT committed:
refactor:adiciona name com user_id

join tabela workouts com tabela users

ElianaMT committed:
refactor:mostra estudent_name quando nao ten treinos cadastrados

ElianaMT committed:
Merge branch 'eliana/revisao/back_3' into sq1-eliana/s07-listagem-treinos-estudante

ElianaMT commented:
Refatora codigo usando nova tabela users_students

ElianaMT commented:
Os ajustes que você sugeriu @CarolineSampaio de try e catch farei ao final. :)

ElianaMT committed:
feat:adiciona os testes unitarios

ElianaMT committed:
feat:adiciona test negativo test_no_workouts_for_unauthenticated_user

ElianaMT committed:
feat:adiciona test negativo

test_student_can_not_get_workouts_from_another_student

ElianaMT committed:
refactor:muda nome do controller a StudentWorkoutController

Apaga Try e Catch

ElianaMT committed:
Merge branch 'eliana/teste/back' into sq1-eliana/s07-listagem-treinos-estudante

ElianaMT commented:
Adiciona testes no backend. 
Elimina Try e Catch no codigo. 
Muda nome do controller a StudentWorkoutController.

CarolineSampaio reviewed (approved):
Testei o funcionamento e está ok, dentro do que você estipulou. Vou realizar a resolução dos conflitos e posterior merge.

CarolineSampaio committed:
Merge branch 'features/admin' of https://github.com/DEVinHouse-Zucchetti/M3P-BackEnd into sq1-eliana/s07-listagem-treinos-estudante
